### ISSUE TITLE
Can you show the source codes of .wasm file?

### ISSUE THREAD

**Issue by DanielLu91** [xibyte/jsketcher#80]
I find you use web assembly for 3d bool operation. If possible, can you show us the source codes for that and show the experience for using web assembly.

**Comment by dumblob** [comment]
I have the same concern. Where from does the file `jsketcher/web/wasm/e0/main.wasm` come from? It doesn't seem to be compiled variant of `jsketcher/web/wasm/e0/main.js`.

**Comment by eric-ma-0709** [comment]
Hi.
first of all, I‘d like to thank you for creating a wonderful javascript projects such as jsketcher.

I analyzed the source code carefully and found that all major operations are embedded in wasm file and compiled from other language using emscripten.
I also found that you were trying to build Brep handling functionality purely in javascript.
It seems to on the way and hasn’t completed yet.

I really want to have access to the source code of the engine so that I can recreate wonderful things from that. If you are going to share it as open source it would be very grateful.
And if you want to make some income from it, I am definitely willing to pay for it.

I hope to hear back from you soon.

**Comment by wolfv** [comment]
You can also check out NoteCAD which has more open source parts available. https://github.com/NoteCAD/NoteCAD

**Comment by eric-ma-0709** [comment]
![image](https://user-images.githubusercontent.com/54482567/73662701-35a2dc00-46d7-11ea-92ef-4036dea29b8a.png)
with this?

And I am looking for a javascript solution.

**Comment by Evil-Spirit** [comment]
@eric-ma-0709
I am afraid, you need to take OpenCASCADE and rewrite it to JavaScript to have BREP in native JavaScript.

**Comment by Evil-Spirit** [comment]
@wolfv 
I have no BREP core at now (just booleans on triangles). Don't know if I will want to do it in future.

**Comment by eric-ma-0709** [comment]
@Evil-Spirit
what do you think about jsketcher? It also has brep handling functionalities.

**Comment by Evil-Spirit** [comment]
As I know, create really robust and precise BREP library is pretty hard task for one-man development. I don't believe if this possible inside pet-projects like jsketcher or notecad.

**Comment by eric-ma-0709** [comment]
@Evil-Spirit 
I have tested some edge cases on jsketcher and it worked well. I think jsketcher is enough for me once its wasm source code is shared. Do you have any idea on where it is from?

**Comment by Evil-Spirit** [comment]
> Do you have any idea on where it is from?

I don't really know (I am not coomon with JS or web, and even not look inside source code of JSketcher), but I suppose, it can be compiled C++ code of third-party geometric core through emscripten.

**Comment by eric-ma-0709** [comment]
@Evil-Spirit
Are there any possibility to bring solvespace to web?

**Comment by Evil-Spirit** [comment]
> Are there any possibility to bring solvespace to web?
Already almost done. All my branches and changes blocked by whitequark (he have no enough time)

**Comment by Evil-Spirit** [comment]
https://files.whitequark.org/SolveSpace/solvespace.html

**Comment by eric-ma-0709** [comment]
@Evil-Spirit 
Where can I see source codes?

**Comment by Evil-Spirit** [comment]
@eric-ma-0709 I think this one branch https://github.com/solvespace/solvespace/tree/emscripten

**Comment by eric-ma-0709** [comment]
@xibyte, could you plz join this thread of forum?

**Comment by mmiscool** [comment]
> Hi.
> first of all, I‘d like to thank you for creating a wonderful javascript projects such as jsketcher.
> 
> I analyzed the source code carefully and found that all major operations are embedded in wasm file and compiled from other language using emscripten.
> I also found that you were trying to build Brep handling functionality purely in javascript.
> It seems to on the way and hasn’t completed yet.
> 
> I really want to have access to the source code of the engine so that I can recreate wonderful things from that. If you are going to share it as open source it would be very grateful.
> And if you want to make some income from it, I am definitely willing to pay for it.
> 
> I hope to hear back from you soon.

@eric-ma-0709 
See #82 

The WASM source code has not had a public release yet and it might be a while before it is. Are you with a company that would want to be using it in a commercial product?

**Comment by xibyte** [comment]
Wasm is a compiled opencascade as well as some c++ supporting code.

**Comment by xibyte** [comment]
I just didn't find any nice way to organize the build yet, but it's on the list for sure. Maybe the build of opencascade to wasm should probably be addressed by even another github project mostly because of its complexity. My main focus so far has been 2d constrainer solver which I am totally rewriting so once I am done with this one I can start looking into 3d engine issues. 

Generally yes, BREP is a big endeavor implementing such can be biting more than I can chew, although jsketcher has it's own native brep engine already implemented with basic boolean operations written in pure javascript and it was being used for some time here. And yes it's not as stable as opancascade.

**Comment by dumblob** [comment]
@xibyte that's very good news! I'd advocate for splitting the opencascade build process to a separate repository.

Btw. what are you using for the 2D solver as backend/engine/algorithms?

**Comment by xibyte** [comment]
I wrote them myself. They are pure client side JavaScript.

**Comment by mmiscool** [comment]
Hello,
We have started working on packaging to make distributing and using the core cad technologies behind JSketcher available to be used in commercial environments. 

See https://www.autodrop3d.com/parametric-cad.html

**Comment by Evil-Spirit** [comment]
Well done, guys!

**Comment by Evil-Spirit** [comment]
But I can't access your link

**Comment by mmiscool** [comment]
Are you talking about https://www.autodrop3d.com/parametric-cad.html link. I just checked it and it seems to be working.

**Comment by Evil-Spirit** [comment]
Sorry, not for me
![www autodrop3d com - Google Chrome 2020-07-21 15 35 20](https://user-images.githubusercontent.com/1980537/88032215-30263a80-cb2d-11ea-8565-aaf4dd6a1d1d.png)

**Comment by mmiscool** [comment]
Looks like the site might be blocked in Russia, Ukraine and parts of India. Don't know why. May need to use a proxy or VPN.

**Comment by mmiscool** [comment]
The build environment for WASM
See https://github.com/Autodrop3d/occt-interpreter

**Comment by dumblob** [comment]
Thank you!

TL;DR pure copy-left with a copyright waiver (it seems there are **no exceptions** - not even for open-source projects so basically you **can not** fork it and maintain your public fork with your own public patch set licensed under some permissive license as MIT)
